File: SimuBoardMac/SimuBoardMac/Models/PointerSound.swift
import Foundation

enum PointerSoundPhase: String, CaseIterable, Sendable {
    case press
    case release
}

enum PointerSoundSample: String, CaseIterable, Sendable {
    case primary = "PRIMARY"
    case secondary = "SECONDARY"
    case middle = "MIDDLE"
}

enum PointerButton: Equatable, Sendable {
    case primary
    case secondary
    case middle
    case auxiliary(Int64)

    var sample: PointerSoundSample {
        switch self {
        case .primary:
            .primary
        case .secondary:
            .secondary
        case .middle, .auxiliary:
            .middle
        }
    }

    /// Adds a subtle distinction when a profile falls back to its primary-button sample.
    var playbackRate: Float {
        switch self {
        case .primary:
            1
        case .secondary:
            0.97
        case .middle:
            1.04
        case .auxiliary:
            1.02
        }
    }

    init(mouseButtonNumber: Int64) {
        self = switch mouseButtonNumber {
        case 0: .primary
        case 1: .secondary
        case 2: .middle
        default: .auxiliary(mouseButtonNumber)
        }
    }
}

struct PointerEvent: Equatable, Sendable {
    let phase: PointerSoundPhase
    let button: PointerButton
}

enum GlobalInputEvent: Sendable {
    case keyboard(KeyboardEvent)
    case pointer(PointerEvent)
}

enum PointerSoundProfile: String, CaseIterable, Identifiable, Sendable {
    case classic
    case silent
    case crisp
    case heavy
    case glass

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .classic: "经典微动"
        case .silent: "静音微动"
        case .crisp: "电竞脆响"
        case .heavy: "厚重办公"
        case .glass: "玻璃触控板"
        }
    }

    var family: String {
        switch self {
        case .classic: "通用鼠标"
        case .silent: "静音鼠标"
        case .crisp: "轻快点击"
        case .heavy: "办公鼠标"
        case .glass: "触控板"
        }
    }

    var tone: String {
        switch self {
        case .classic: "清晰、均衡"
        case .silent: "柔和、低调"
        case .crisp: "短促、明亮"
        case .heavy: "低沉、扎实"
        case .glass: "干净、通透"
        }
    }
}

File: BattutaWindows/src/Battuta.Core/Audio/PointerSound.cs
namespace Battuta.Core.Audio;

public enum PointerSoundPhase
{
    Press,
    Release,
}

public enum PointerSoundSample
{
    Primary,
    Secondary,
    Middle,
}

public enum PointerButtonKind
{
    Primary,
    Secondary,
    Middle,
    Auxiliary,
}

public readonly record struct PointerButton
{
    private PointerButton(PointerButtonKind kind, long auxiliaryNumber)
    {
        Kind = kind;
        AuxiliaryNumber = auxiliaryNumber;
    }

    public PointerButtonKind Kind { get; }
    public long AuxiliaryNumber { get; }

    public static PointerButton Primary => new(PointerButtonKind.Primary, 0);
    public static PointerButton Secondary => new(PointerButtonKind.Secondary, 1);
    public static PointerButton Middle => new(PointerButtonKind.Middle, 2);
    public static PointerButton Auxiliary(long number) => new(PointerButtonKind.Auxiliary, number);

    public static PointerButton FromButtonNumber(long number) => number switch
    {
        0 => Primary,
        1 => Secondary,
        2 => Middle,
        _ => Auxiliary(number),
    };

    public PointerSoundSample Sample => Kind switch
    {
        PointerButtonKind.Primary => PointerSoundSample.Primary,
        PointerButtonKind.Secondary => PointerSoundSample.Secondary,
        PointerButtonKind.Middle or PointerButtonKind.Auxiliary => PointerSoundSample.Middle,
        _ => PointerSoundSample.Primary,
    };

    public float PlaybackRate => Kind switch
    {
        PointerButtonKind.Primary => 1,
        PointerButtonKind.Secondary => 0.97f,
        PointerButtonKind.Middle => 1.04f,
        PointerButtonKind.Auxiliary => 1.02f,
        _ => 1,
    };
}

public readonly record struct PointerSoundProfileId
{
    public PointerSoundProfileId(string value)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(value);
        Value = value;
    }

    public string Value { get; }
    public override string ToString() => Value ?? string.Empty;
}

public sealed record PointerSoundProfileDefinition(
    PointerSoundProfileId Id,
    string DisplayName,
    string Family,
    string Tone);

public static class PointerSoundProfiles
{
    public static readonly PointerSoundProfileId Classic = new("classic");
    public static readonly PointerSoundProfileId Silent = new("silent");
    public static readonly PointerSoundProfileId Crisp = new("crisp");
    public static readonly PointerSoundProfileId Heavy = new("heavy");
    public static readonly PointerSoundProfileId Glass = new("glass");
}

public static class PointerSoundProfileCatalog
{
    public static IReadOnlyList<PointerSoundProfileDefinition> All { get; } =
    [
        new(PointerSoundProfiles.Classic, "经典微动", "通用鼠标", "清晰、均衡"),
        new(PointerSoundProfiles.Silent, "静音微动", "静音鼠标", "柔和、低调"),
        new(PointerSoundProfiles.Crisp, "电竞脆响", "轻快点击", "短促、明亮"),
        new(PointerSoundProfiles.Heavy, "厚重办公", "办公鼠标", "低沉、扎实"),
        new(PointerSoundProfiles.Glass, "玻璃触控板", "触控板", "干净、通透"),
    ];

    private static readonly Dictionary<string, PointerSoundProfileDefinition> ById =
        All.ToDictionary(profile => profile.Id.Value, StringComparer.Ordinal);

    public static PointerSoundProfileDefinition Default => Get(PointerSoundProfiles.Classic);

    public static bool TryGet(string? id, out PointerSoundProfileDefinition profile)
    {
        if (id is not null && ById.TryGetValue(id, out profile!))
        {
            return true;
        }

        profile = null!;
        return false;
    }

    public static PointerSoundProfileDefinition Get(PointerSoundProfileId id) =>
        TryGet(id.Value, out var profile)
            ? profile
            : throw new KeyNotFoundException($"Unknown pointer profile: {id.Value}");
}
